File: 2019/day_one/tyrannyOfTheRocketEquation.ts
/*
The Typescript solution for: https://adventofcode.com/2019/day/1
The answer to part one is: 3311492
*/ 
import * as fs from 'node:fs';

function readFileFunction(): Promise<string> {
    // read and return the input file text
    return new Promise((resolve, reject) => {
        fs.readFile('input.txt', 'utf-8', (err, data) => {
            if(err) {
                console.error('cannot read local file')
                reject('cannot read local file')
            }
            resolve(data)
        });
    }
)}


function calculateEquation(listVelocities: Array<string>): number{
    /*
    Given a list of numbers, calculate the fuel required by 
    dividing its value by three, rounding down, and 
    then subtracting 2
    */
    let totalFuelRequired = 0;

    for(let velocity of listVelocities){
        let tempVelocity: number = (velocity as unknown) as number;

        totalFuelRequired += Math.floor(tempVelocity / 3) - 2
    }

    return totalFuelRequired;
}

async function main(): Promise<string> {
    /*
        Read the input file, and call associated functions to perfrom calculation for 
        the fuel required.
    */
    let fileData: string;

    try {
        fileData = await readFileFunction();
    } catch(err) {
        console.error('error reading file');
        return('error while trying to read file');
    }

    let velocities: Array<string> = fileData.split('\n');
    const totalFuelNeeded: number = calculateEquation(velocities);

    return `The answer to part one is: ${totalFuelNeeded}`; 
}

// run code and print the result to screen
let result: string = await main();
console.log(result);

File: 2019/day_one/tyrannyOfTheRocketEquation.py
"""
Day one advent of code in python - https://adventofcode.com/2019/day/1

the answer for part one is: 3311492
"""
import math 

def find_mass(inputlist):
    """
    Given a list of numbers, calculate the fuel required by 
    dividing its value by three, rounding down, and 
    then subtracting 2
    """
    total = 0

    for i in inputlist: 
        total += (math.floor(i/3) - 2)

    return total

if __name__ == "__main__": 
    file_input = []
    with open('input.txt', 'r') as inputFile: 
        file_input = [int(line.strip()) for line in inputFile]

    total_mass = find_mass(file_input)

    print(f"the answer for part one is: {total_mass}")